### PR TITLE
Add imperative invoker updates to popover API

### PR DESCRIPTION
**Pull request by lukewarlow**
### Description

Add imperative invoker updates to popover API

showPopover and togglePopover can now take an options parameter.

### Motivation

So people can discover this new functionality.

### Additional details

Released in Chromium 133 - https://chromestatus.com/feature/5120638407409664

### Related issues and pull requests

Relates to https://github.com/mdn/mdn/issues/598 (this doesn't cover the anchor positioning change)

### CODE REVIEW

**Comment by lukewarlow** [comment]
Probably needs a browser compat data update too.

**Comment by lukewarlow** [comment]
cc @chrisdavidmills I just noticed you mentioned working on this, so feel free to close this if you've already done it or want more detailed changes!

**Comment by hamishwillee** [comment]
@lukewarlow Howdy, is there an associated BCD update showing this is implemented anywhere?

File: files/en-us/web/api/htmlelement/showpopover/index.md [hunk 1]
```diff
@@ -15,12 +15,16 @@ When `showPopover()` is called on an element with the [`popover`](/en-US/docs/We
...
 ```
 
 ### Parameters
 
-None.
+- `options` {{optional_inline}}
+  - : An object with the following properties:
+    - `source` {{optional_inline}}
+      - An {{domxref("HTMLElement")}} that triggered the popover, if any. This provides all the same functionality that using
+        [`popovertarget`](/en-US/docs/Web/HTML/Element/button#popovertarget) would provide.
```
Comment (hamishwillee) [inline files/en-us/web/api/htmlelement/showpopover/index.md:27]:
```suggestion
      - An {{domxref("HTMLElement")}} that triggered the popover, if any.
        This provides all the same functionality that using [`popovertarget`](/en-US/docs/Web/HTML/Element/button#popovertarget) would provide.
```

File: files/en-us/web/api/htmlelement/togglepopover/index.md [hunk 2]
```diff
@@ -22,14 +22,24 @@ When `togglePopover()` is called on an element with the [`popover`](/en-US/docs/
...
     - If set to `false`, the popover is hidden if it was initially shown. If it was initially hidden, nothing happens.
+- `options` {{optional_inline}}
+  - : An object with the following properties:
+    - `force` {{optional_inline}}
+      - : A boolean, which causes `togglePopover()` to behave like {{domxref("HTMLElement.showPopover", "showPopover()")}} or {{domxref("HTMLElement.hidePopover", "hidePopover()")}}, except that it doesn't throw an exception if the popover is already in the target state.
+        - If set to `true`, the popover is shown if it was initially hidden. If it was initially shown, nothing happens.
+        - If set to `false`, the popover is hidden if it was initially shown. If it was initially hidden, nothing happens.
+    - `source` {{optional_inline}}
+      - An {{domxref("HTMLElement")}} that triggered the popover, if any. This provides all the same functionality that using
+        [`popovertarget`](/en-US/docs/Web/HTML/Element/button#popovertarget) would provide.
```
Comment (hamishwillee) [inline files/en-us/web/api/htmlelement/togglepopover/index.md:42]:
```suggestion
      - An {{domxref("HTMLElement")}} that triggered the popover, if any.
        This provides all the same functionality that using [`popovertarget`](/en-US/docs/Web/HTML/Element/button#popovertarget) would provide.
```

File: files/en-us/web/api/htmlelement/showpopover/index.md [hunk 3]
```diff
@@ -10,17 +10,22 @@ browser-compat: api.HTMLElement.showPopover
...
 
-When `showPopover()` is called on an element with the [`popover`](/en-US/docs/Web/HTML/Global_attributes/popover) attribute that is currently hidden, a {{domxref("HTMLElement/beforetoggle_event", "beforetoggle")}} event will be fired, followed by the popover showing, and then the {{domxref("HTMLElement/toggle_event", "toggle")}} event firing. If the element is already showing, an error will be thrown.
+When `showPopover()` is called on an element with the [`popover`](/en-US/docs/Web/HTML/Global_attributes/popover) attribute that is currently hidden, a {{domxref("HTMLElement/beforetoggle_event", "beforetoggle")}} event will be fired, followed by the popover showing, and then the {{domxref("HTMLElement/toggle_event", "toggle")}} event firing.
+If the element is already showing, an error will be thrown.
 
 ## Syntax
 
 ```js-nolint
-showPopover()
+showPopover(options)
```
Comment (hamishwillee) [inline files/en-us/web/api/htmlelement/showpopover/index.md:19]:
```suggestion
showPopover()
showPopover(options)
```

File: files/en-us/web/api/htmlelement/showpopover/index.md [hunk 4]
```diff
@@ -10,17 +10,23 @@ browser-compat: api.HTMLElement.showPopover
...
 ```
 
 ### Parameters
 
-None.
+- `options` {{optional_inline}}
+  - : An object with the following properties:
+    - `source` {{optional_inline}}
+      - : An {{domxref("HTMLElement")}} that triggered the popover, if any.
+        This provides all the same functionality that using [`popovertarget`](/en-US/docs/Web/HTML/Element/button#popovertarget) would provide.
```
Comment (hamishwillee) [inline files/en-us/web/api/htmlelement/showpopover/index.md:29]: Sorry, but I don't understand what this is for, why it is needed, or how it is used.

Say I have a button and I trigger the popover with that button, why does the popover care what triggered it. What if I lie or omit this? Seems to make zero difference if it is defined or not.

I get `popovertarget` - that creates a link from a button (say) to the popover, but in code you create that link by ... calling this method on the thing you want want to show. 

Can you create an example in both the docs that shows how this is used, and explain those things?
Also would be good to add an example in https://developer.mozilla.org/en-US/docs/Web/API/Popover_API/Using. If this is somehow like `popovertarget`, perhaps drawing the connection would be good.

**Comment by chrisdavidmills** [comment]
> cc @chrisdavidmills I just noticed you mentioned working on this, so feel free to close this if you've already done it or want more detailed changes!

Thanks @lukewarlow. I was holding off on working on this until https://github.com/whatwg/html/pull/10728 was merged into the spec, but I see it has now happened. I can handle the other bits later in the month, or early next month.

File: files/en-us/web/api/htmlelement/showpopover/index.md (hunk 4 above)
Comment (hamishwillee) [reply files/en-us/web/api/htmlelement/showpopover/index.md:29]: @lukewarlow @chrisdavidmills Either way, this can't merge as is because IMO no one can understand the point of this without doing extra reading.

Comment (lukewarlow) [reply files/en-us/web/api/htmlelement/showpopover/index.md:29]: Apologies I meant to respond here, while coming up with the more detailed explanation I actually found a bug in chrome's implementation so I was double checking that my understanding was correct. I'll spend some time to make this more detailed.

Fwiw here's the 4 things this does:

- It correctly handles nested popover scenarios, if you invoke a popover from within a popover it shouldn't close it. Currently that only works declaratively, this new feature lets you do it imperatively.

- It fixes up the focus order such that the contents of the popover are the next tab stop once opened. Currently this only works declaratively. This new feature will do this imperatively (this is the bug I found).

- It might have an impact on light dismiss behaviour. (This might be both a spec and implementation bug).

- Both imperative and declarative now setup an implicit anchor element for anchor positioning (this patch doesn't cover that I'd prefer if that was a follow-up)

Comment (hamishwillee) [reply files/en-us/web/api/htmlelement/showpopover/index.md:29]: Thanks @lukewarlow - looking forward to seeing what you come up with -  FWIW IMO a live example is likely to be needed. 

I'm around this week but away in first 2-ish weeks of January and sporadically in between. You're welcome to seek another reviewer if I'm not around when you get to this.

**Comment by lukewarlow** [comment]
I'm going to go ahead and close this as I don't think I can concisely explain the complexity so will leave it for someone else to do.

TLDR:

With the new options the popover gets the same behaviour as it does from popovertarget buttons:

1. When the "source" element is nested inside a popover, this new popover *won't* close the parent popover, allowing nested popover behaviour.
2. The "source" element becomes an implicit anchor element for the popover, which makes anchor positioning easier to utilitise (this is also a novel addition for popovertarget buttons).
3. The popover element is moved to be next in the focus order from the "source" element (This isn't actually implemented in Chromium currently due to a bug in the implementation of the source option but it is an important aspect of the feature).

**Comment by chrisdavidmills** [comment]
No worries @lukewarlow. I ended up covering these new features in this PR: https://github.com/mdn/content/pull/37990. I'll look at your content to make sure I'm not missing anything.
